pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,484 @@
+		}
+
+		delay := time.NewTimer(maxRetryTimeout)
+		select {
+		case <-delay.C:
+			continue
+		case <-ctx.Done():
+			if !delay.Stop() {
+				<-delay.C
+			}
+			return nil, fmt.Errorf("%s opening stream to %s: context closed", t.ownID, dest)
+		}
+	}
+	return nil, fmt.Errorf("%s failed to open stream to %s: %w", t.ownID, dest, err)
+}
+
+func (t *Transport) sendPayload(dest types.NodeID, payload []byte) error {
+	t.nodesLock.RLock()
+	defer t.nodesLock.RUnlock()
+
+	// There are two cases when we get an error:
+	// 1. We don't have the node ID in the nodes table. E.g. we didn't call Connect().
+	// 2. We created an entry for the node but have not opened a connection.
+	// But if we added the node via Connect, then it should open the connection.
+	// It doesn't make sense to open a new one here.
+	stream, err := t.getNodeStreamAndInfoWithoutLock(dest)
+	if err != nil {
+		return err
+	}
+
+	w := bufio.NewWriter(stream)
+	if _, err = w.Write(payload); err != nil {
+		return err
+	}
+	return w.Flush()
+}
+
+func (t *Transport) encode(msg *messagepb.Message) ([]byte, error) {
+	p, err := proto.Marshal(msg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal message: %w", err)
+	}
+
+	tm := TransportMessage{t.ownID.Pb(), p}
+	buf := new(bytes.Buffer)
+	if err = tm.MarshalCBOR(buf); err != nil {
+		return nil, fmt.Errorf("failed to CBOR marshal message: %w", err)
+	}
+	return buf.Bytes(), nil
+}
+
+func (t *Transport) decode(s io.Reader) (*messagepb.Message, types.NodeID, error) {
+	var tm TransportMessage
+	err := tm.UnmarshalCBOR(s)
+	if err != nil {
+		return nil, "", err
+	}
+
+	var msg messagepb.Message
+	if err := proto.Unmarshal(tm.Payload, &msg); err != nil {
+		return nil, "", err
+	}
+	return &msg, types.NodeID(tm.Sender), nil
+}
+
+func (t *Transport) Send(ctx context.Context, dest types.NodeID, msg *messagepb.Message) error {
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	out, err := t.encode(msg)
+	if err != nil {
+		return err
+	}
+
+	err = t.sendPayload(dest, out)
+	if err != nil {
+		t.reconnect(ctx, dest)
+		return errors.Wrapf(err, "%s failed to send data to %s", t.ownID, dest)
+	}
+
+	return nil
+}
+
+func (t *Transport) mirHandler(s network.Stream) {
+	t.logger.Log(logging.LevelDebug, "mir handler started", "src", t.ownID, "dst", s.ID())
+	defer t.logger.Log(logging.LevelDebug, "mir handler stopped", "src", t.ownID, "dst", s.ID())
+
+	defer func() {
+		t.logger.Log(logging.LevelDebug, "mir handler is closing stream", "src", t.ownID, "dst", s.ID())
+		err := s.Close()
+		if err != nil {
+			t.logger.Log(logging.LevelError, "closing stream", "src", t.ownID, "dst", s.ID(), "err", err)
+		}
+	}() // nolint
+
+	for {
+		msg, sender, err := t.decode(s)
+		if err != nil {
+			t.logger.Log(logging.LevelError, "failed to read mir request", "src", t.ownID, "sender", sender, "err", err)
+			return
+		}
+		t.incomingMessages <- events.ListOf(
+			events.MessageReceived(types.ModuleID(msg.DestModule), sender, msg),
```

Comment:
I don't think we can trust `sender` here, since we receive it from the other, potentially faulty, node and doesn't seem to verify in any way.

Reply:
I can do the following:
```
                msg, sender, err := t.decode(s)
		if err != nil {
			t.logger.Log(logging.LevelError, "failed to read mir request", "src", t.ownID, "sender", sender, "err", err)
			return
		}
		t.nodesLock.RLock()
		v, ok := t.nodes[sender]
		t.nodesLock.RUnlock()
		if !ok {
			return
		}
		msgSenderID := v.AddrInfo.ID.String()
		senderPeerID := s.Conn().RemotePeer().String()
		if msgSenderID != senderPeerID {
			t.logger.Log(logging.LevelWarn, 
				"failed to validate sender",
				"src", t.ownID, "sender", sender, "msgSender", msgSenderID, "peerID", senderPeerID)
			return
		}
		t.incomingMessages <- events.ListOf(
			events.MessageReceived(types.ModuleID(msg.DestModule), sender, msg),
		)
```

Reply:
Fixed

Reply:
Now we have some additional overhead for every received message. We may want to improve this in future.